src/commands/kv/bucket/upload.rs

```diff
@@ -9,24 +9,98 @@ use crate::settings::global_user::GlobalUser;
 use crate::settings::target::Target;
 use crate::terminal::message;
 
+const KEY_MAX_SIZE: usize = 512;
+const VALUE_MAX_SIZE: usize = 2 * 1024 * 1024;
+const PAIRS_MAX_COUNT: usize = 10000;
+const UPLOAD_MAX_SIZE: usize = 100 * 1024 * 1024;
+
 pub fn upload(
     target: &Target,
     user: GlobalUser,
     namespace_id: &str,
     path: &Path,
 ) -> Result<(), failure::Error> {
-    let pairs: Result<Vec<KeyValuePair>, failure::Error> = match &metadata(path) {
+    let mut pairs: Vec<KeyValuePair> = match &metadata(path) {
         Ok(file_type) if file_type.is_dir() => directory_keys_values(path),
         Ok(_file_type) => {
             // any other file types (files, symlinks)
             failure::bail!("wrangler kv:bucket upload takes a directory")
         }
         Err(e) => failure::bail!("{}", e),
-    };
+    }?;
+
+    validate_file_uploads(pairs.clone())?;
+
+    // Iterate over all key-value pairs and create batches of uploads, each of which are
+    // maximum 10K key-value pairs in size OR maximum ~50MB in size. Upload each batch
+    // as it is created.
+    let mut key_count = 0;
+    let mut key_pair_bytes = 0;
+    let mut key_value_batch: Vec<KeyValuePair> = Vec::new();
+
+    while !(pairs.is_empty() && key_value_batch.is_empty()) {
+        if pairs.is_empty() {
+            // Last batch to upload
+            call_put_bulk_api(target, user.clone(), namespace_id, &mut key_value_batch)?;
+        } else {
+            let pair = pairs.pop().unwrap();
+            if key_count + pair.key.len() > PAIRS_MAX_COUNT
+            // Keep upload size small to keep KV bulk API happy
+            || key_pair_bytes + pair.key.len() + pair.value.len() > UPLOAD_MAX_SIZE * 1/2
+            {
+                call_put_bulk_api(target, user.clone(), namespace_id, &mut key_value_batch)?;
```

Comment:
what do we do when this errors?

Reply:
the error is bubbled up and printed out using our default error handling logic

Reply:
(i've seen this in action many times, especially when i was figuring out how to navigate timeouts!)

Reply:
i guess more what i mean is, how do we rectify a partial upload? if we are hashing contents, i think we're mostly fine, but we might want to consider doing the `delete` part of `sync` _after_ the `upload` part of `sync` to prevent leaving users in a weird half state, as that will at least leave their old assets up.

Reply:
Ah, this makes sense!